admittance_controller/include/admittance_controller/admittance_controller.hpp

```diff
@@ -0,0 +1,157 @@
+  using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;
+
+  class AdmittanceController : public controller_interface::ChainableControllerInterface {
+  public:
+    ADMITTANCE_CONTROLLER_PUBLIC
+    AdmittanceController() {
+      int o =0;
+    };
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    CallbackReturn on_init() override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    controller_interface::InterfaceConfiguration command_interface_configuration() const override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    controller_interface::InterfaceConfiguration state_interface_configuration() const override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    CallbackReturn on_configure(const rclcpp_lifecycle::State &previous_state) override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    CallbackReturn on_activate(const rclcpp_lifecycle::State &previous_state) override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    CallbackReturn on_deactivate(const rclcpp_lifecycle::State &previous_state) override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    CallbackReturn on_cleanup(const rclcpp_lifecycle::State &previous_state) override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    CallbackReturn on_error(const rclcpp_lifecycle::State &previous_state) override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    controller_interface::return_type update_and_write_commands(
+        const rclcpp::Time &time, const rclcpp::Duration &period) override;
+
+
+  protected:
+    std::vector<hardware_interface::CommandInterface> on_export_reference_interfaces() override;
+
+    controller_interface::return_type update_reference_from_subscribers() override;
+
+    bool on_set_chained_mode(bool chained_mode) override;
+
+    int num_joints_ = 0;
+
+    // vectors to hold joint hardware interfaces
+//    std::vector<hardware_interface::LoanedCommandInterface> joint_position_command_interface_;
+//    std::vector<hardware_interface::LoanedCommandInterface> joint_velocity_command_interface_;
+//    std::vector<hardware_interface::LoanedCommandInterface> joint_acceleration_command_interface_;
+//    std::vector<hardware_interface::LoanedCommandInterface> joint_effort_command_interface_;
+//    std::vector<hardware_interface::LoanedStateInterface> joint_position_state_interface_;
+//    std::vector<hardware_interface::LoanedStateInterface> joint_velocity_state_interface_;
+//    std::vector<hardware_interface::LoanedStateInterface> joint_acceleration_state_interface_;
+//    std::vector<hardware_interface::LoanedStateInterface> joint_effort_state_interface_;
+    size_t state_pos_ind = -1;
+    size_t state_vel_ind = -1;
+    size_t state_acc_ind = -1;
+    size_t command_pos_ind = -1;
+    size_t command_vel_ind = -1;
+    size_t command_acc_ind = -1;
+
+
+    // internal reference values
+    const std::vector<std::string> reference_interfaces_types_ = {"position", "velocity"};
+    std::vector<double*> position_reference_;
+    std::vector<double*> velocity_reference_;
```

Comment:
A double* is a red flag to me. Are you sure you need a pointer here? Can it just be a reference (&)? This is unusual.

Reply:
I changed this to use a reference wrapper